src/analog/private/m2kanalogin_impl.cpp

```diff
@@ -42,14 +42,14 @@ using namespace std::placeholders;
 
 class M2kAnalogIn::M2kAnalogInImpl : public DeviceIn {
 public:
-	M2kAnalogInImpl(iio_context * ctx, std::string adc_dev, bool sync) :
+	M2kAnalogInImpl(iio_context * ctx, std::string adc_dev, bool sync, M2kHardwareTrigger *trigger) :
 		DeviceIn(ctx, adc_dev),
 		m_need_processing(false),
 		m_trigger(nullptr)
 	{
 		m_m2k_fabric = make_shared<DeviceGeneric>(ctx, "m2k-fabric");
 		m_ad5625_dev = std::make_shared<DeviceGeneric>(ctx, "ad5625");
-		m_trigger = new M2kHardwareTrigger(ctx);
+		m_trigger =  trigger;
```

Comment:
you're setting the m_trigger twice. First was in the initializer list.
What happens if someone passes nullptr for trigger? M2kAnalogInImpl will work with limited functionality? or crash? should this constructor throw?

Reply:
You're right, something is missing here.
How I see things: one way to go would be to throw an exception here: https://github.com/analogdevicesinc/libm2k/pull/54/files#diff-b7d2236a13d61e896bfdcd1529c58fddR66 . An M2K should always have a valid trigger structure, if the object is not correctly created then the M2K Context should not work at all (so we throw the exception right at the beginning).

Any thoughts on this? @dNechita @adisuciu ?

Reply:
Changed to throw an exception where it first creates the trigger object .